File: esque-core/src/test/java/org/loesak/esque/core/EsqueIT.java
package org.loesak.esque.core;

import org.elasticsearch.client.Request;
import org.elasticsearch.client.Response;
import org.elasticsearch.client.RestClient;
import org.junit.jupiter.api.Test;
import org.loesak.esque.core.elasticsearch.RestClientOperations;
import org.loesak.esque.core.elasticsearch.documents.MigrationRecord;

import java.io.IOException;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class EsqueIT extends AbstractElasticsearchIT {

    @Test
    void execute_createsEsqueIndex() throws IOException {
        try (RestClient client = createRestClient()) {
            Esque esque = new Esque(client, "create-index-test");
            esque.execute();

            Response response = client.performRequest(new Request("HEAD", "/.esque"));
            assertThat(response.getStatusLine().getStatusCode()).isEqualTo(200);
        }
    }

    @Test
    void execute_runsAllMigrations() throws IOException {
        try (RestClient client = createRestClient()) {
            Esque esque = new Esque(client, "run-all-test");
            esque.execute();

            // verify all three test indices were created
            assertIndexExists(client, "/test-index-v1");
            assertIndexExists(client, "/test-index-v2");
            assertIndexExists(client, "/test-index-v3");
        }
    }

    @Test
    void execute_recordsMigrationHistory() throws IOException {
        String migrationKey = "history-test";
        try (RestClient client = createRestClient()) {
            Esque esque = new Esque(client, migrationKey);
            esque.execute();

            RestClientOperations ops = new RestClientOperations(client, migrationKey);
            List<MigrationRecord> records = ops.getMigrationRecords();

            assertThat(records).hasSize(3);

            // verify order
            assertThat(records.get(0).order()).isEqualTo(0);
            assertThat(records.get(0).filename()).isEqualTo("V1.0.0__CreateTestIndex.yml");
            assertThat(records.get(0).version()).isEqualTo("1.0.0");
            assertThat(records.get(0).description()).isEqualTo("CreateTestIndex");

            assertThat(records.get(1).order()).isEqualTo(1);
            assertThat(records.get(1).filename()).isEqualTo("V1.1.0__CreateSecondIndex.yml");

            assertThat(records.get(2).order()).isEqualTo(2);
            assertThat(records.get(2).filename()).isEqualTo("V2.0.0__CreateThirdIndex.yml");

            // verify metadata is populated
            for (MigrationRecord record : records) {
                assertThat(record.migrationKey()).isEqualTo(migrationKey);
                assertThat(record.checksum()).isNotNull();
                assertThat(record.installedOn()).isNotNull();
                assertThat(record.executionTime()).isNotNull();
                assertThat(record.executionTime()).isGreaterThanOrEqualTo(0L);
            }
        }
    }

    @Test
    void execute_isIdempotent() throws IOException {
        String migrationKey = "idempotent-test";
        try (RestClient client = createRestClient()) {
            // first execution
            new Esque(client, migrationKey).execute();

            RestClientOperations ops = new RestClientOperations(client, migrationKey);
            List<MigrationRecord> firstRun = ops.getMigrationRecords();
            assertThat(firstRun).hasSize(3);

            // second execution should not fail and should not create duplicate records
            new Esque(client, migrationKey).execute();

            List<MigrationRecord> secondRun = ops.getMigrationRecords();
            assertThat(secondRun).hasSize(3);

            // records should be identical
            for (int i = 0; i < 3; i++) {
                assertThat(secondRun.get(i).filename()).isEqualTo(firstRun.get(i).filename());
                assertThat(secondRun.get(i).checksum()).isEqualTo(firstRun.get(i).checksum());
                assertThat(secondRun.get(i).installedOn()).isEqualTo(firstRun.get(i).installedOn());
            }
        }
    }

    @Test
    void execute_differentKeysAreIndependent() throws IOException {
        try (RestClient client = createRestClient()) {
            String key1 = "independent-key-1";
            String key2 = "independent-key-2";

            // execute migrations under key1
            new Esque(client, key1).execute();

            RestClientOperations ops1 = new RestClientOperations(client, key1);
            RestClientOperations ops2 = new RestClientOperations(client, key2);

            // key1 should have 3 migration records
            List<MigrationRecord> records1 = ops1.getMigrationRecords();
            assertThat(records1).hasSize(3);
            records1.forEach(r -> assertThat(r.migrationKey()).isEqualTo(key1));

            // key2 should have no records - migrations are scoped to the key
            List<MigrationRecord> records2 = ops2.getMigrationRecords();
            assertThat(records2).isEmpty();
        }
    }

    @Test
    void execute_recordsUserWhenProvided() throws IOException {
        String migrationKey = "user-test";
        String user = "integration-test-user";
        try (RestClient client = createRestClient()) {
            Esque esque = new Esque(client, migrationKey, user);
            esque.execute();

            RestClientOperations ops = new RestClientOperations(client, migrationKey);
            List<MigrationRecord> records = ops.getMigrationRecords();

            for (MigrationRecord record : records) {
                assertThat(record.installedBy()).isEqualTo(user);
            }
        }
    }

    @Test
    void execute_recordsNullUserWhenNotProvided() throws IOException {
        String migrationKey = "no-user-test";
        try (RestClient client = createRestClient()) {
            Esque esque = new Esque(client, migrationKey);
            esque.execute();

            RestClientOperations ops = new RestClientOperations(client, migrationKey);
            List<MigrationRecord> records = ops.getMigrationRecords();

            for (MigrationRecord record : records) {
                assertThat(record.installedBy()).isNull();
            }
        }
    }

    private void assertIndexExists(RestClient client, String indexPath) throws IOException {
        Response response = client.performRequest(new Request("HEAD", indexPath));
        assertThat(response.getStatusLine().getStatusCode()).isEqualTo(200);
    }
}

File: esque-core/src/test/kotlin/org/loesak/esque/core/EsqueIT.kt
package org.loesak.esque.core

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.elasticsearch.client.Request
import org.junit.jupiter.api.Test
import org.loesak.esque.core.elasticsearch.RestClientOperations

class EsqueIT : AbstractElasticsearchIT() {

  private val templateProperties = mapOf("templatedIndexName" to "test-index-v4")

  @Test
  fun execute_createsEsqueIndex() {
    createRestClient().use { client ->
      Esque(client, EsqueConfiguration(migrationKey = "create-index-test"), templateProperties)
          .execute()
      val response = client.performRequest(Request("HEAD", "/.esque"))
      assertThat(response.statusLine.statusCode).isEqualTo(200)
    }
  }

  @Test
  fun execute_runsAllMigrations() {
    createRestClient().use { client ->
      Esque(client, EsqueConfiguration(migrationKey = "run-all-test"), templateProperties).execute()
      assertIndexExists(client, "/test-index-v1")
      assertIndexExists(client, "/test-index-v2")
      assertIndexExists(client, "/test-index-v3")
      assertIndexExists(client, "/test-index-v4")
    }
  }

  @Test
  fun execute_recordsMigrationHistory() {
    val migrationKey = "history-test"
    createRestClient().use { client ->
      Esque(client, EsqueConfiguration(migrationKey = migrationKey), templateProperties).execute()

      val records = RestClientOperations(client, migrationKey).getMigrationRecords()
      assertThat(records).hasSize(4)

      assertThat(records[0].order).isEqualTo(0)
      assertThat(records[0].filename).isEqualTo("V1.0.0__CreateTestIndex.yml")
      assertThat(records[0].version).isEqualTo("1.0.0")
      assertThat(records[0].description).isEqualTo("CreateTestIndex")

      assertThat(records[1].order).isEqualTo(1)
      assertThat(records[1].filename).isEqualTo("V1.1.0__CreateSecondIndex.yml")

      assertThat(records[2].order).isEqualTo(2)
      assertThat(records[2].filename).isEqualTo("V2.0.0__CreateThirdIndex.yml")

      assertThat(records[3].order).isEqualTo(3)
      assertThat(records[3].filename).isEqualTo("V3.0.0__CreateTemplatedIndex.yml")

      for (record in records) {
        assertThat(record.migrationKey).isEqualTo(migrationKey)
        assertThat(record.checksum).isNotNull()
        assertThat(record.installedOn).isNotNull()
        assertThat(record.executionTime).isNotNull()
        assertThat(record.executionTime).isGreaterThanOrEqualTo(0L)
      }
    }
  }

  @Test
  fun execute_isIdempotent() {
    val migrationKey = "idempotent-test"
    createRestClient().use { client ->
      Esque(client, EsqueConfiguration(migrationKey = migrationKey), templateProperties).execute()
      val firstRun = RestClientOperations(client, migrationKey).getMigrationRecords()
      assertThat(firstRun).hasSize(4)

      Esque(client, EsqueConfiguration(migrationKey = migrationKey), templateProperties).execute()
      val secondRun = RestClientOperations(client, migrationKey).getMigrationRecords()
      assertThat(secondRun).hasSize(4)

      for (i in 0..3) {
        assertThat(secondRun[i].filename).isEqualTo(firstRun[i].filename)
        assertThat(secondRun[i].checksum).isEqualTo(firstRun[i].checksum)
        assertThat(secondRun[i].installedOn).isEqualTo(firstRun[i].installedOn)
      }
    }
  }

  @Test
  fun execute_differentKeysAreIndependent() {
    createRestClient().use { client ->
      val key1 = "independent-key-1"
      val key2 = "independent-key-2"

      Esque(client, EsqueConfiguration(migrationKey = key1), templateProperties).execute()

      val records1 = RestClientOperations(client, key1).getMigrationRecords()
      assertThat(records1).hasSize(4)
      records1.forEach { assertThat(it.migrationKey).isEqualTo(key1) }

      val records2 = RestClientOperations(client, key2).getMigrationRecords()
      assertThat(records2).isEmpty()
    }
  }

  @Test
  fun execute_recordsUserWhenProvided() {
    val migrationKey = "user-test"
    val user = "integration-test-user"
    createRestClient().use { client ->
      Esque(
              client,
              EsqueConfiguration(migrationKey = migrationKey, migrationUser = user),
              templateProperties,
          )
          .execute()
      val records = RestClientOperations(client, migrationKey).getMigrationRecords()
      for (record in records) {
        assertThat(record.installedBy).isEqualTo(user)
      }
    }
  }

  @Test
  fun execute_recordsNullUserWhenNotProvided() {
    val migrationKey = "no-user-test"
    createRestClient().use { client ->
      Esque(client, EsqueConfiguration(migrationKey = migrationKey), templateProperties).execute()
      val records = RestClientOperations(client, migrationKey).getMigrationRecords()
      for (record in records) {
        assertThat(record.installedBy).isNull()
      }
    }
  }

  @Test
  fun execute_withPropertiesAndNoPlaceholdersInFiles_succeedsAndIgnoresExtraProperties() {
    createRestClient().use { client ->
      Esque(
              client,
              EsqueConfiguration(migrationKey = "extra-properties-test"),
              mapOf("unused" to "value", "templatedIndexName" to "test-index-v4"),
          )
          .execute()
      assertIndexExists(client, "/test-index-v1")
    }
  }

  @Test
  fun execute_withTemplateVariables_substitutesValuesInPathAndCreatesCorrectIndex() {
    createRestClient().use { client ->
      Esque(
              client,
              EsqueConfiguration(migrationKey = "templating-substitution-test"),
              templateProperties,
          )
          .execute()
      assertIndexExists(client, "/test-index-v4")
    }
  }

  @Test
  fun execute_withMissingTemplateVariable_throwsBeforeAnyMigrationRuns() {
    createRestClient().use { client ->
      assertThatThrownBy {
            Esque(client, EsqueConfiguration(migrationKey = "missing-var-test")).execute()
          }
          .rootCause()
          .hasMessageContaining("templatedIndexName")

      val records = RestClientOperations(client, "missing-var-test").getMigrationRecords()
      assertThat(records).isEmpty()
    }
  }

  private fun assertIndexExists(client: org.elasticsearch.client.RestClient, indexPath: String) {
    val response = client.performRequest(Request("HEAD", indexPath))
    assertThat(response.statusLine.statusCode).isEqualTo(200)
  }
}
